test/unit/kubernetes-deploy/kubernetes_resource/custom_resource_definition_test.rb

```diff
@@ -0,0 +1,200 @@
+    assert(crd.validation_failed?, "Missing failure_conditions key should fail validation")
+    assert_equal(crd.validation_error_msg, "Missing required top-level key(s): [:failure_conditions]\n" \
+      "success_conditions must contain at least one entry")
+  end
+
+  def test_rollout_conditions_raises_error_with_invalid_json
+    crd = build_crd(merge_rollout_annotation('bad string'))
+    crd.validate_definition(kubectl)
+    assert(crd.validation_failed?, "Invalid rollout conditions were accepted")
+    assert(crd.validation_error_msg.match(/Error parsing rollout conditions/))
+  end
+
+  def test_cr_instance_fails_validation_when_rollout_conditions_for_crd_invalid
+    crd = build_crd(merge_rollout_annotation('bad string'))
+    cr = KubernetesDeploy::KubernetesResource.build(namespace: "test", context: "test",
+      logger: @logger, statsd_tags: @statsd_tags, crd: crd,
+      definition: {
+        "kind" => "UnitTest",
+        "metadata" => { "name" => "test" },
+      })
+    cr.validate_definition(kubectl)
+    assert(crd.validation_error_msg.match(/Error parsing rollout conditions/))
+  end
+
+  def test_cr_instance_valid_when_rollout_conditions_for_crd_valid
+    rollout_conditions = {
+      success_conditions: [
+        {
+          path: "$.test_path",
+          value: "test_value",
+        },
+      ],
+      failure_conditions: [
+        {
+          path: "$.test_path",
+          value: "test_value",
+        },
+      ],
+    }.to_json
+
+    crd = build_crd(merge_rollout_annotation(rollout_conditions))
+    cr = KubernetesDeploy::KubernetesResource.build(namespace: "test", context: "test",
+      logger: @logger, statsd_tags: @statsd_tags, crd: crd,
+      definition: {
+        "kind" => "UnitTest",
+        "metadata" => { "name" => "test" },
+      })
+    cr.validate_definition(kubectl)
+    refute(cr.validation_failed?)
+  end
+
+  def test_instance_timeout_annotation
+    crd = build_crd(crd_spec.merge(
+      "metadata" => {
+        "name" => "unittests.stable.example.io",
+      },
+    ))
+    cr = KubernetesDeploy::KubernetesResource.build(namespace: "test", context: "test",
+      logger: @logger, statsd_tags: @statsd_tags, crd: crd,
+      definition: { "kind" => "UnitTest", "metadata" => { "name" => "test" } })
+    assert_equal(cr.timeout, KubernetesDeploy::CustomResource.timeout)
+
+    crd = build_crd(crd_spec.merge(
+      "metadata" => {
+        "name" => "unittests.stable.example.io",
+        "annotations" => {
+          KubernetesDeploy::CustomResourceDefinition::TIMEOUT_FOR_INSTANCE_ANNOTATION => "60S",
+        },
+      }
+    ))
+    cr = KubernetesDeploy::KubernetesResource.build(namespace: "test", context: "test",
+      logger: @logger, statsd_tags: @statsd_tags, crd: crd,
+      definition: { "kind" => "UnitTest", "metadata" => { "name" => "test" } })
+    assert_equal(cr.timeout, 60)
+  end
+
+  def test_instance_timeout_messages_with_rollout_conditions
```

Comment:
Do the timeout tests actually take the full timeout duration? If so should we do something to speed that up?